overlord/devicestate/devicemgr.go

```diff
@@ -329,22 +333,79 @@ func (m *DeviceManager) maybeSetupUbuntuSave() error {
 	if !runMntSaveMounted {
 		// we don't have ubuntu-save, save will be used directly
 		logger.Noticef("no ubuntu-save mount")
-		m.saveAvailable = true
 		return nil
 	}
 
-	logger.Noticef("bind-mounting ubuntu-save under %v", dirs.SnapSaveDir)
+	sysd := systemd.New(systemd.SystemMode, progress.Null)
+
+	// In newer core20/core22 we have a mount unit for ubuntu-save, which we
+	// will try to start first. Invoking systemd-mount in this case would fail.
+	err = sysd.Start([]string{"var-lib-snapd-save.mount"})
+	if err == nil {
+		logger.Noticef("mount unit for ubuntu-save was started")
+		return nil
+	} else {
+		// We only fall through and mount directly if the failure was because of a missing
+		// mount file, which possible does not exist. Any other failure we treat as an actual
+		// error.
+		if !strings.Contains(err.Error(), "Unit var-lib-snapd-save.mount not found.") {
+			return err
+		}
+	}
 
-	err = systemd.New(systemd.SystemMode, progress.Null).Mount(boot.InitramfsUbuntuSaveDir,
-		dirs.SnapSaveDir, "-o", "bind")
+	// Otherwise try to directly mount the partition with systemd-mount.
+	logger.Noticef("bind-mounting ubuntu-save under %v", dirs.SnapSaveDir)
+	err = sysd.Mount(boot.InitramfsUbuntuSaveDir, dirs.SnapSaveDir, "-o", "bind")
 	if err != nil {
 		logger.Noticef("bind-mounting ubuntu-save failed %v", err)
 		return fmt.Errorf("cannot bind mount %v under %v: %v", boot.InitramfsUbuntuSaveDir, dirs.SnapSaveDir, err)
 	}
-	m.saveAvailable = true
 	return nil
 }
 
+// ensureUbuntuSaveSnapFolders creates the necessary folder structure for
+// /var/lib/snapd/save/snap/<snap>. This is normally done during installation
+// of a snap, but there are two cases where this can be insufficient.
+//
+// 1. When migrating to a newer snapd, folders are not automatically created for
+//    snaps that are already installed. They will only be created during a refresh of
+//    the snap itself, whereas we want to cover all the cases.
+// 2. During install mode for the gadget/kernel/etc, the folders are not created.
+//    So this function can be invoked as a part of system-setup.
+func (m *DeviceManager) ensureUbuntuSaveSnapFolders() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	snaps, err := snapstate.All(m.state)
+	if err != nil {
+		return err
+	}
+
+	for _, s := range snaps {
+		saveDir := snap.CommonDataSaveDir(s.InstanceName())
+		if err := os.MkdirAll(saveDir, 0755); err != nil {
```

Comment:
this code seems to be reached also in the unit tests, but I don't see an assertion about created directories

Reply:
Good point! Added a new unit test that verifies these folders are created during install